File: leetCode.java/15.3-sum.java
import java.util.LinkedList;
import java.util.List;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

/*
 * @lc app=leetcode id=15 lang=java
 *
 * [15] 3Sum
 *
 * https://leetcode.com/problems/3sum/description/
 *
 * algorithms
 * Medium (25.92%)
 * Likes:    7207
 * Dislikes: 830
 * Total Accepted:    953.2K
 * Total Submissions: 3.6M
 * Testcase Example:  '[-1,0,1,2,-1,-4]'
 *
 * Given an array nums of n integers, are there elements a, b, c in nums such
 * that a + b + c = 0? Find all unique triplets in the array which gives the
 * sum of zero.
 * 
 * Note:
 * 
 * The solution set must not contain duplicate triplets.
 * 
 * Example:
 * 
 * 
 * Given array nums = [-1, 0, 1, 2, -1, -4],
 * 
 * A solution set is:
 * [
 * ⁠ [-1, 0, 1],
 * ⁠ [-1, -1, 2]
 * ]
 * 
 * 
 */

// @lc code=start
class Solution {
    // testcase:[-1, 0, 1, 2, -1, -4, -1, -1, -1, -1]
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new LinkedList<>();
        if (nums == null || nums.length < 3) {
            return res;
        }

        Arrays.sort(nums);

        for (int i = 0; i < nums.length - 2; i++) {
            if (i != 0 && nums[i - 1] == nums[i]) {
                continue;
            }
            int a = nums[i];
            if (a > 0) {
                break;
            }
            int left = i + 1;
            int right = nums.length - 1;
            while (left < right) {
                int b = nums[left];
                int c = nums[right];
                if (a + b > 0) {
                    break;
                }
                if (a + b + c == 0) {
                    res.add(Arrays.asList(a, b, c));
                    left++;
                    right--;
                    while (left < right && nums[left] == nums[left - 1]) {
                        left++;
                    }
                    while (left < right && nums[right] == nums[right + 1]) {
                        right--;
                    }
                } else if (a + b + c < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return res;

        // 313/313 cases passed (25 ms)
        // Your runtime beats 68.16 % of java submissions
        // Your memory usage beats 9.12 % of java submissions (56.6 MB)

    }

}
// @lc code=end

File: leetCode.python/15.3-sum.py
#
# @lc app=leetcode id=15 lang=python3
#
# [15] 3Sum
#
from typing import *

# @lc code=start
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        # print(f"nums is {nums}")
        # 三元数组顺序不重要，i<j<k即可
        # 时间复杂度O(n^2) ->每一个数都要进行n的操作
        # 空间复杂度O(1)
        ans = []
        for i in range(len(nums) - 2):
            if (i > 0 and nums[i] == nums[i-1]):
                continue
            if (nums[i] + nums[i + 1] + nums[i + 2] > 0):
                break # 因为如果加最前面的都大于0那肯定不用再试了，i之后的数字也不用再试了
            if (nums[i] + nums[-1] + nums[-2] < 0):
                continue # nums[i]肯定是凑不到0了，看下一个数了
            two_sum = self.twoSum(nums[i+1:], -nums[i])
            if two_sum:
                ans.extend([nums[i]] + j for j in two_sum)
        return ans

    def twoSum(self, nums: List[int], target: int) -> List[List[int]]:
        l, r = 0, len(nums) - 1
        ans = []
        while l < r:
            if (nums[l] + nums[r] > target):
                r -= 1
            elif (nums[l] + nums[r] < target):
                l += 1
            else:
                ans.append([nums[l], nums[r]])
                l += 1
                # r -= 1
                # 不对，因为可能有多个重复的，所以l和r要到完全不一样为止
                while l < r and nums[l] == nums[l - 1]:
                    l += 1
                r -= 1
                while l < r and nums[r] == nums[r + 1]:
                    r -= 1 
        return ans

        
# @lc code=end

